pkg/virt-controller/watch/vm.go

```diff
@@ -1673,6 +1673,56 @@ func (c *VMController) removeVMIFinalizer(vmi *virtv1.VirtualMachineInstance) er
 	return err
 }
 
+func (c *VMController) removeVMFinalizer(vm *virtv1.VirtualMachine) (*virtv1.VirtualMachine, error) {
+	vmCopy := vm.DeepCopy()
+	controller.RemoveFinalizer(vmCopy, virtv1.VirtualMachineControllerFinalizer)
+
+	if equality.Semantic.DeepEqual(vm.Finalizers, vmCopy.Finalizers) {
+		return vm, nil
+	}
+
+	log.Log.V(3).Object(vm).Infof("VM is in a final state. Removing VM controller finalizer")
+
+	ops := []string{}
+	newFinalizers, err := json.Marshal(vmCopy.Finalizers)
+	if err != nil {
+		return vm, err
+	}
+	oldFinalizers, err := json.Marshal(vm.Finalizers)
+	if err != nil {
+		return vm, err
+	}
+	ops = append(ops, fmt.Sprintf(`{ "op": "test", "path": "/metadata/finalizers", "value": %s }`, string(oldFinalizers)))
+	ops = append(ops, fmt.Sprintf(`{ "op": "replace", "path": "/metadata/finalizers", "value": %s }`, string(newFinalizers)))
+	vm, err = c.clientset.VirtualMachine(vm.Namespace).Patch(vm.Name, types.JSONPatchType, controller.GeneratePatchBytes(ops), &v1.PatchOptions{})
+	return vm, err
+}
+
+func (c *VMController) addVMFinalizer(vm *virtv1.VirtualMachine) (*virtv1.VirtualMachine, error) {
+	vmCopy := vm.DeepCopy()
+	controller.AddFinalizer(vmCopy, virtv1.VirtualMachineControllerFinalizer)
```

Comment:
:+1:  Thanks! For the moment I extended the `add` and `remove` functions taking the finalizer you want to delete as parameter. This will help future extensions.

Reply:
Awesome! Thanks!